param_sync.py

```diff
@@ -7,12 +7,39 @@ class ParamSyncRule:
     training.
     """
 
+    def update_core(self, local_params, master_params):
+        """
+        Iterator over updates for local and master params.
+
+        This function must be ready to deal with symbolic or numerical
+        inputs.
```

Comment:
I wanted to leave the option of doing the computation on the CPU as before.  Unless there is a compelling reason not to, then we can remove this requirement and also update_params below.